File: server/src/main/com/broll/gainea/server/core/objects/MapObject.java
package com.broll.gainea.server.core.objects;

import com.broll.gainea.net.NT_BoardObject;
import com.broll.gainea.server.core.GameContainer;
import com.broll.gainea.server.core.map.Location;
import com.broll.gainea.server.core.map.Ship;
import com.broll.gainea.server.core.objects.buffs.BuffableInt;
import com.broll.gainea.server.core.player.Player;
import com.broll.gainea.server.core.processing.GameUpdateReceiverAdapter;

public abstract class MapObject extends GameUpdateReceiverAdapter {
    protected GameContainer game;

    protected Player owner;

    private Location location;

    private int id;

    private int icon = 0;

    private String name;

    private float scale = 1;

    protected int moveCount;

    protected BuffableInt<MapObject> movesPerTurn = new BuffableInt<>(this, 1); //default 1 move

    public void init(GameContainer game) {
        this.id = game.newObjectId();
        this.game = game;
    }

    public void turnStart() {
        moveCount = 0;
    }

    public boolean hasRemainingMove() {
        return moveCount < movesPerTurn.getValue();
    }

    public void moved() {
        moveCount++;
    }

    public BuffableInt<MapObject> getMovesPerTurn() {
        return movesPerTurn;
    }

    public Location getLocation() {
        return location;
    }

    public MapObject() {
        this(null);
    }

    public MapObject(Player owner) {
        this.owner = owner;
    }

    public void setLocation(Location location) {
        this.location = location;
    }

    public void setScale(float scale) {
        this.scale = scale;
    }

    public String getName() {
        return name;
    }

    public void setName(String name) {
        this.name = name;
    }

    public float getScale() {
        return scale;
    }

    public int getId() {
        return id;
    }

    public void setIcon(int icon) {
        this.icon = icon;
    }

    public int getIcon() {
        return icon;
    }

    public boolean canMoveTo(Location to) {
        if (!to.isTraversable()) {
            return false;
        }
        if (to instanceof Ship) {
            if (!((Ship) to).passable(location)) {
                return false;
            }
        }
        if (location instanceof Ship) {
            if (((Ship) location).getTo() != to) {
                return false;
            }
        }
        return true;
    }

    public NT_BoardObject nt() {
        NT_BoardObject object = new NT_BoardObject();
        fillObject(object);
        return object;
    }

    protected void fillObject(NT_BoardObject object) {
        object.id = (short) id;
        object.name = name;
        object.icon = (short) icon;
        object.size = scale * 30;
        if (location != null) {
            object.location = (short) location.getNumber();
        }
    }

    public Player getOwner() {
        return owner;
    }

    @Override
    public String toString() {
        return "MapObject{" +
                "id=" + id +
                ", name='" + name +
                ", location=" + location + '\'' +
                '}';
    }
}

File: server/src/main/com/broll/gainea/server/core/objects/MapObject.kt
package com.broll.gainea.server.core.objects

import com.broll.gainea.net.NT_BoardObject
import com.broll.gainea.server.core.Game
import com.broll.gainea.server.core.map.Location
import com.broll.gainea.server.core.map.Ship
import com.broll.gainea.server.core.map.SpawnLocation
import com.broll.gainea.server.core.objects.buffs.BuffableInt
import com.broll.gainea.server.core.player.Player
import com.broll.gainea.server.core.player.isNeutral
import com.broll.gainea.server.core.processing.GameUpdateReceiverAdapter

abstract class MapObject(var owner: Player) : GameUpdateReceiverAdapter() {
    protected lateinit var game: Game
    var location: Location = SpawnLocation
    var id = 0
        private set
    var icon = 0
    var name: String? = null
    var scale = 1f
    var description: String? = null
    protected var moveCount = 0
    var movesPerTurn = BuffableInt(this, 1) //default 1 move

    open fun init(game: Game) {
        id = game.newObjectId()
        this.game = game
    }

    open fun prepareForTurnStart() {
        moveCount = 0
    }

    open fun hasRemainingMove() = moveCount < movesPerTurn.value


    open fun moved(fromPlayerAction: Boolean = false) {
        if (fromPlayerAction) {
            moveCount++
        }
    }

    open fun canMoveTo(to: Location): Boolean {
        if (!to.traversable) {
            return false
        }
        if (to is Ship) {
            if (!to.passable(location)) {
                return false
            }
        }
        if (location is Ship) {
            if ((location as Ship).to !== to) {
                return false
            }
        }
        return true
    }

    val moveTargets: List<Location>
        get() = location.connectedLocations.filter { canMoveTo(it) }

    open fun nt(): NT_BoardObject {
        val obj = NT_BoardObject()
        fillObject(obj)
        return obj
    }

    protected fun fillObject(obj: NT_BoardObject) {
        obj.id = id.toShort()
        obj.name = name
        obj.icon = icon.toShort()
        obj.scale = scale
        obj.description = description
        if (!owner.isNeutral()) {
            obj.owner = owner.serverPlayer.id.toShort()
        }
        if (location != SpawnLocation) {
            obj.location = location.number.toShort()
        }
    }

    override fun toString(): String {
        return "MapObject{" +
                "id=" + id +
                ", name='" + name +
                ", location=" + location + '\'' +
                '}'
    }
}
